### DISCUSSION TITLE
derived async gets stale value (previously `loadable` seems to tear)

### DISCUSSION THREAD

**Discussion by dmaskasky** [pmndrs/jotai#2192] [Bug report]
# derived async gets stale value

Issue https://github.com/pmndrs/jotai/issues/2191

## Example:
https://codesandbox.io/s/jovial-kepler-d9v27h

### Steps to reproduce: 
  (1) click the increment button multiple times in short succession.

### Expected
The final count is displayed (debounced by the async delay)

### Actual
The async count value can lag behind the real count.

**Comment by dai-shi** [comment]
I'm not sure if this is an issue with `loadable` itself.
Can you reproduce the issue without `atomEffect`?

`unsub` looks weird, and it may read a value from a stale closure, but removing it doesn't change the behavior as well as abort controller.
Here's a simplified version: https://codesandbox.io/s/cocky-turing-n6jndk?file=/src/Counter.tsx

**Reply by dai-shi** [reply]
Seems like the issue is reproducible without `loadable`.
https://codesandbox.io/s/relaxed-swirles-tsxvfc?file=/src/Counter.tsx
(I added `aborted` variable, which is not important for the repro.)

**Reply by dmaskasky** [reply]
hmm, I can still get your example to tear.

https://github.com/pmndrs/jotai/assets/5034873/0bca14a3-8225-43ad-b1f1-cdb609e6fd08

**Reply by dmaskasky** [reply]
I'm looking into whether this might be an issue with atomEffect.

**Comment by dmaskasky** [comment]
alright, I've narrowed this problem down to get(asyncAtom) not being reliably added to the jotai atom dependencies.

**Reply by dai-shi** [reply]
Unfortunately, it's not too surprising with a possible bug related with promise handling. Like #2168
Can you try if Jotai v2.2.3 works for your case?

**Reply by dmaskasky** [reply]
same issue with 2.2.3 as well.
Here is the issue without atomEffect.

https://codesandbox.io/s/jovial-kepler-d9v27h?file=/src/App.jsx

**Reply by dmaskasky** [reply]
The `await void 0` seems to make the dependency collection unstable.

**Reply by dmaskasky** [reply]
Interestingly, if you await a timeout 1000, it works again. I wonder if the evaluation of the dependency map is raced with the asyncAtom promise.

**Reply by dai-shi** [reply]
https://codesandbox.io/s/jovial-grass-sd8stc
If it waits for a tick, it works. I don't think it ever worked before.
This seems like a buggy behavior, but not sure it's fixable (it might break some other tests).

You are very welcome to dig into it further,
but for now, I think the workaround for `atomEffect` is:
```diff
-   await void 0
+   await new Promise((r) => setTimeout(r)) // wait for a tick
```

**Reply by dai-shi** [reply]
I don't know why but changing to this works:
```js
const defer = () => Promise.resolve();
```
https://codesandbox.io/s/ancient-leftpad-3k6z9d?file=/src/App.jsx

Maybe a bundler/transpiler thing?

**Reply by dmaskasky** [reply]
https://github.com/pmndrs/jotai/assets/5034873/a601770b-2c06-453d-8a6b-c4513c29ed37

i've been able to repro in both csb and cra

**Reply by dai-shi** [reply]
Do you mean that you can reproduce the behavior with `await delay(10)` instead of `await delay(0)`?
That sounds very bad...

**Reply by dai-shi** [reply]
https://codesandbox.io/s/busy-darkness-wjs5kj
Ah, with `await delay(10)`, it reproduces.

So, it was my mistake in https://github.com/pmndrs/jotai/discussions/2192#discussioncomment-7309786. My bad.

**Comment by dmaskasky** [comment]
That doesn’t work either.

**Reply by dai-shi** [reply]
That's unexpected, because the csb repro seems to work. 🤔

**Reply by dai-shi** [reply]
Do you mean it doesn't work even if we wait 1 sec?
```js
await new Promise((r) => setTimeout(r, 1000))
```

**Reply by dmaskasky** [reply]
Oh, its not changing the dependencyMap its reusing old promises when its not supposed to.

```js
const derivedAsyncAtom = atom(
  async (get, { setSelf }) => {
    const count = get(countAtom);
    await defer();
    console.log(`derived (${count})`);
    const content = await get(asyncAtom);
    console.log(`derived (${count})`, content);
    setSelf(contentAtom, content);
  },
  (_get, set, ...args) => set(...args)
);
```

```
derived (7)
derived (8)
derived (9)
derived (7) here we go! 🚀 (7)
derived (9) here we go! 🚀 (7)
derived (8) here we go! 🚀 (7)
```

**Reply by dmaskasky** [reply]
The promises don't look to be resolving in sequence. Hmm.

**Reply by dmaskasky** [reply]
A button increments count 5 times before the first promise resolves.
Next, the first promise resolves. This resolved value is cached and is used for all 5 increment events. `<-- this is incorrect`
The first promise `(7)` resolve triggers a rerun of derived atom read.
Then, immediately the remaining 4 increment events trigger derived atom read.
Then, the remaining 4 promises resolve in order, and do not trigger derived atom read.

```
derived (7)
derived (8)
derived (9)
derived (10)
derived (11)
resolved (7)
derived (7) here we go! 🚀 (7)
derived (11) here we go! 🚀 (7)
derived (10) here we go! 🚀 (7)
derived (9) here we go! 🚀 (7)
derived (8) here we go! 🚀 (7)
resolved (8)
resolved (9)
resolved (10)
resolved (11)
```

```js
const asyncAtom = atom(async (get) => {
  const id = get(countAtom);
  await delay(2000);
  console.log(`resolved (${id})`);
  return `here we go! 🚀 (${id})`;
});

const derivedAsyncAtom = atom(async (get) => {
  const count = get(countAtom);
  await defer();
  console.log(`derived (${count})`);
  const content = await get(asyncAtom);
  console.log(`derived (${count})`, content);
});
```

**Reply by dai-shi** [reply]
Alright, it seems like a bug.
Can you please open a new issue with the simplified codesandbox reproduction (without loadable, without atomEffect, as few lines as possible)?
Our first task would be to create a failing test.

**Reply by dai-shi** [reply]
Related: https://github.com/pmndrs/jotai/discussions/2195#discussioncomment-7322232

**Comment by dmaskasky** [comment]
This test fails.

https://codesandbox.io/s/jovial-kepler-d9v27h?file=/src/App.test.js

```js
import { renderHook, act } from "@testing-library/react";
import { atom, useAtom, useAtomValue } from "jotai";

console.clear();

test.only("resolves dependencies reliably after a delay", async () => {
  const countAtom = atom(0);
  const asyncCountAtom = atom(0);

  const resolve = [];
  const asyncAtom = atom(async (get) => {
    const count = get(countAtom);
    await new Promise((r) => resolve.push(r));
    console.log(`resolved (${count})`);
    return count;
  });

  const derivedAsyncAtom = atom(
    async (get, { setSelf }) => {
      const count = get(countAtom);
      await microtask();
      console.log(`derived (${count})`);
      const asyncCount = await get(asyncAtom);
      console.log(`derived (${count})`, asyncCount);
      setSelf(asyncCountAtom, asyncCount);
    },
    (_get, set, ...args) => set(...args)
  );

  const derivedSyncAtom = atom((get) => {
    get(derivedAsyncAtom);
  });

  function useTest() {
    const asyncCount = useAtomValue(asyncCountAtom);
    useAtom(derivedSyncAtom);
    const [count, setCount] = useAtom(countAtom);
    return { count, setCount, asyncCount };
  }

  const { result } = renderHook(useTest);
  const { setCount } = result.current;

  await microtask();

  expect(result.current.count).toBe(0);
  expect(result.current.asyncCount).toBe(0);

  await act(() => setCount(increment));
  await act(() => setCount(increment));

  resolve[0]();
  resolve[1]();
  resolve[2]();

  await microtask();
  await microtask();
  await microtask();
  await macrotask();

  expect(result.current.count).toBe(2);
  expect(result.current.asyncCount).toBe(2);

  await act(() => setCount(increment));
  await act(() => setCount(increment));

  resolve[3]();
  resolve[4]();

  await microtask();
  await microtask();
  await microtask();
  await macrotask();

  expect(result.current.count).toBe(4);
  expect(result.current.asyncCount).toBe(4); // 3
});

function macrotask() {
  return new Promise((r) => setTimeout(r, 0));
}

async function microtask() {}

function increment(c) {
  return c + 1;
}
```

```derived (0)
derived (1)
derived (2)
resolved (0)
resolved (1)
resolved (2)
derived (2) 2
derived (1) 2
derived (0) 2
derived (3)
derived (4)
resolved (3)
resolved (4)
derived (3) 3
derived (4) 3
```

**Reply by dai-shi** [reply]
Nice. I think we can create a test with `render` instead of `renderHook`.

**Reply by dmaskasky** [reply]
Add tests with PR https://github.com/pmndrs/jotai/pull/2199
